test/test_stone.py

```diff
@@ -1315,6 +1445,340 @@ def test_forward_reference(self):
+            "to a pre-existing data_type.", cm.exception.msg)
+        self.assertEqual(cm.exception.lineno, 10)
+
+        # Try patching type with pre-existing name of different type
+        text = textwrap.dedent("""\
+            namespace files
+            struct QuotaInfo
+                "The space quota info for a user."
+                user_id String
+                    "The user associated with the quota."
+                example default
+                    user_id="1234"
+                example pro
+                    user_id="1234P"
+            patch union QuotaInfo
+                quota UInt64
+                    "The user's total quota allocation (bytes)."
+            """)
+        with self.assertRaises(InvalidSpec) as cm:
+            api = specs_to_ir([('test.stone', text)])
+
+        self.assertEqual("Type mismatch. Patch 'QuotaInfo' corresponds to " +
+            "pre-existing data_type 'QuotaInfo' (test.stone:2) that has " +
+            "type other than 'union'.", cm.exception.msg)
+        self.assertEqual(cm.exception.lineno, 10)
+
+        # Try patching field with pre-existing name
+        text = textwrap.dedent("""\
+            namespace files
+            struct QuotaInfo
+                "The space quota info for a user."
+                user_id String
+                    "The user associated with the quota."
+                example default
+                    user_id="1234"
+                example pro
+                    user_id="1234P"
+            patch struct QuotaInfo
+                user_id Int32
+                    "The user associated with the quota."
+            """)
+        with self.assertRaises(InvalidSpec) as cm:
+            api = specs_to_ir([('test.stone', text)])
+
+        self.assertEqual("Patched field 'user_id' overrides " +
+            "pre-existing field in 'QuotaInfo' (test.stone:4).", cm.exception.msg)
+        self.assertEqual(cm.exception.lineno, 11)
+
+        # Try patching field with example tag that doesn't exist
+        text = textwrap.dedent("""\
+            namespace files
+            struct QuotaInfo
+                "The space quota info for a user."
+                user_id String
+                    "The user associated with the quota."
+                example default
+                    user_id="1234"
+                example pro
+                    user_id="1234P"
+            patch struct QuotaInfo
+                quota UInt64
+                    "The user associated with the quota."
+                example doesNotExist
+                    user_id="1234"
+            """)
+        with self.assertRaises(InvalidSpec) as cm:
+            api = specs_to_ir([('test.stone', text)])
+
+        self.assertEqual("Example defined in patch 'QuotaInfo' must " +
+            "correspond to a pre-existing example.", cm.exception.msg)
+        self.assertEqual(cm.exception.lineno, 13)
+
+    def test_union_patch_semantics(self):
```

Comment:
Make sure you have a test for patching an open union with a closed union and vice versa.

Reply:
Done.